inference-engine/src/mkldnn_plugin/nodes/space_to_batch.cpp

```diff
@@ -21,200 +21,285 @@ class SpaceToBatchImpl: public ExtLayerBase {
-        const size_t OD = layout == NCDHW ? outDims[dims_size - 3] : 1lu;
-        const size_t OH = outDims[dims_size - 2];
-        const size_t OW = outDims[dims_size - 1];
-
-        const int64_t cBSD = layout == NCDHW ? _block_shape[dims_size - 3] : 1lu;  // Do not use name BSD. It affects MacOS build
-        const int64_t BSH = _block_shape[dims_size - 2];
-        const int64_t BSW = _block_shape[dims_size - 1];
-
-        const int64_t PF = layout == NCDHW ? _pads_begin[dims_size - 3] : 0;
-        const int64_t PT = _pads_begin[dims_size - 2];
-        const int64_t PL = _pads_begin[dims_size - 1];
-
-        const size_t OH_OW = OH * OW;
-        const size_t IH_IW = IH * IW;
-        const size_t ID_IH_IW = ID * IH_IW;
-        const size_t IC_ID_IH_IW = IC * ID_IH_IW;
-
-        const size_t work_amount = OB*OC*OD*OH*OW;
-
-        auto thread_body = [&](const int ithr, const int nthr) {
-            size_t start(0lu), end(0lu);
-            splitter(work_amount, nthr, ithr, start, end);
-            if (start >= end)
-                return;
-            int64_t ob(0), oc(0), od(0), oh(0), ow(0);
-            parallel_it_init(start, ob, OB, oc, OC, od, OD, oh, OH, ow, OW);
-
-            for (; ob < OB; ob++) {
-                const int64_t ib = ob % IB;
-                const int64_t ib_k = ib * IC_ID_IH_IW;
-                int64_t bi = ob / IB;
-                const int64_t shift_w = bi % BSW - PL;
-                bi /= BSW;
-                const int64_t shift_h = (layout == NCDHW ? bi % BSH : bi) - PT;
-                const int64_t shift_d = layout == NCDHW ? (bi / BSH - PF) : 0;
-                for (; oc < OC; oc++) {
-                    const int64_t ic_k = ib_k + oc * ID_IH_IW;
-                    for (; od < OD; od++) {
-                        const int64_t id = od * cBSD + shift_d;
-                        if (id < 0 || id >= ID) {
-                            std::fill(dst_data + start, dst_data + start + OH_OW, T(0));
-                            start += OH_OW;
-                            if (start >= end)
-                                break;
-                            continue;
+        auto inShape5D  = getShape5D(outDims);
+        auto outShape5D = getShape5D(inDims);
+
+        size_t inSpatialStep = inShape5D[2] * inShape5D[3] * inShape5D[4];
+        size_t inBatchStep   = inShape5D[1] * inSpatialStep;
+
+        size_t outSpatialStep = outShape5D[2] * outShape5D[3] * outShape5D[4];
+        size_t outBatchStep   = outShape5D[1] * outSpatialStep;
+
+        const int64_t padFront = padsBegin[dimsSize - 3];
+        const int64_t padTop   = padsBegin[dimsSize - 2];
+        const int64_t padLeft  = padsBegin[dimsSize - 1];
+
+        size_t ID = dimsSize == 5 ? inShape5D[dimsSize - 3] : 1lu;
+        size_t OD = dimsSize == 5 ? outShape5D[dimsSize - 3] : 1lu;
+
+        if (layout == NHWC || layout == NDHWC) {
+            parallel_for(inShape5D[0], [&](size_t i0) {
```

Comment:
I don't sure that using `parallel_for` by batch is optimized. I think that SpaceToBatch has probably usually small batch (specific of this layer?). Maybe I'm wrong.

Reply:
We discussed offline. It's ok because line 129:
`auto inShape5D  = getShape5D(outDims);`